zcash_keys/src/keys/transparent.rs

```diff
@@ -0,0 +1,340 @@
+    fn from(value: bip32::Error) -> Self {
+        ParseError::Bip32(value)
+    }
+}
+
+/// A secp256k1 secret key, along with a flag indicating whether a compressed encoding should be
+/// used when performing DER serialization.
+pub struct Key {
+    secret: SecretKey,
+    compressed: bool,
+}
+
+impl Key {
+    /// Decodes a base58-encoded secret key.
+    ///
+    /// This corresponds to <https://github.com/zcash/zcash/blob/1f1f7a385adc048154e7f25a3a0de76f3658ca09/src/key_io.cpp#L282>
+    pub fn decode_base58<N: NetworkConstants>(
+        network: &N,
+        encoded: &SecretString,
+    ) -> Result<Self, ParseError> {
+        let decoded = SecretVec::new(
+            bs58::decode(encoded.expose_secret())
+                .with_check(None)
+                .into_vec()?,
+        );
+        let prefix = network.b58_secret_key_prefix();
+        let decoded_len = decoded.expose_secret().len();
+        let compressed = decoded_len == (33 + prefix.len());
+        if (decoded_len == 32 + prefix.len()
+            || (compressed && decoded.expose_secret().last() == Some(&1)))
+            && decoded.expose_secret()[0..prefix.len()] == prefix
+        {
+            let key_end = decoded_len - if compressed { 1 } else { 0 };
+            let bytes = PrivateKeyBytes::try_from(&decoded.expose_secret()[prefix.len()..key_end])?;
+            Ok(Self {
+                secret: SecretKey::from_bytes(&bytes)?,
+                compressed,
+            })
+        } else {
+            Err(ParseError::LengthInvalid)
+        }
+    }
+
+    /// Decodes a base58-encoded secret key.
+    ///
+    /// This corresponds to <https://github.com/zcash/zcash/blob/1f1f7a385adc048154e7f25a3a0de76f3658ca09/src/key_io.cpp#L298>
+    pub fn encode_base58<N: NetworkConstants>(&self, network: &N) -> SecretString {
+        let input = SecretVec::new(
+            network
+                .b58_secret_key_prefix()
+                .iter()
+                .chain(self.secret.secret_bytes().iter())
+                .chain(if self.compressed { &[1] } else { &[0] })
+                .copied()
+                .collect(),
+        );
+
+        SecretString::new(
+            bs58::encode(input.expose_secret())
+                .with_check()
+                .into_string(),
+        )
+    }
+
+    /// Returns the wrapped secp256k1 secret key.
+    pub fn secret(&self) -> SecretKey {
+        self.secret
+    }
+
+    /// Returns the value of the compressed flag.
+    pub fn compressed(&self) -> bool {
+        self.compressed
+    }
+
+    /// Derives the secp256k1 public key corresponding to the secret key.
+    pub fn pubkey(&self) -> PublicKey {
+        let secp = secp256k1::Secp256k1::new();
+        self.pubkey_internal(&secp)
+    }
+
+    /// Derives the secp256k1 public key corresponding to the secret key,
+    /// using the provided secp context.
+    pub fn pubkey_internal<C: Signing>(&self, secp: &Secp256k1<C>) -> PublicKey {
```

Comment:
Why is a public API named `*_internal`? Would be better as e.g. `pubkey_with_context` or `pubkey_with`.